### ISSUE TITLE
Keyboard cursor should go to end of line when TextField is set via TextEditController

### ISSUE THREAD

**Issue by cornelius-tan** [flutter/flutter#11416]
## Steps to Reproduce

The keyboard cursor should go to the end of line when TextField is set via TextEditController (for example when autocomplete), so that users can continue typing from the cursor position. Instead the cursor always RESET to the beginning of line when the text is set.

1. Run the code below.
2. Type 'Flu'
3. Tap the FAB button to set text to 'Flutter' (simulate autocomplete).
4. Expect cursor to jump to the end of line. Actual: Cursor reset to beginning of line.

## Code

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(new MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      title: 'Flutter Demo',
      theme: new ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: new MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => new _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final TextEditingController _controller = new TextEditingController();

  @override
  Widget build(BuildContext context) {
    return new Scaffold(
      appBar: new AppBar(
        title: new TextField(
          controller: _controller,
          autofocus: true,
        )
      ),
      body: new Container(),
      floatingActionButton: new FloatingActionButton(
        onPressed: () {
          _controller.text = 'Flutter';
        },
        tooltip: 'Autocomplete',
        child: new Icon(Icons.add),
      ),
    );
  }
}
```

## Flutter Doctor

```
[√] Flutter (on Microsoft Windows [Version 10.0.15063], locale en-US, channel master)
    • Flutter at C:\Users\tzm\Downloads\flutter_sdk
    • Framework revision 1d8705f3db (2 hours ago), 2017-07-26 11:44:50 -0700
    • Engine revision 3a12bc092d
    • Tools Dart version 1.25.0-dev.7.0

[√] Android toolchain - develop for Android devices (Android SDK 25.0.3)
    • Android SDK at C:\Users\tzm\AppData\Local\Android\sdk
    • Platform android-26, build-tools 25.0.3
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[√] Android Studio (version 2.3)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[√] IntelliJ IDEA Community Edition (version 2017.1)
    • Flutter plugin version 15.2
    • Dart plugin version 171.4424

[√] Connected devices
    • SM G7102                  • 772e5ea4      • android-arm • Android 4.4.2 (API 19)
    • Android SDK built for x86 • emulator-5554 • android-x86 • Android 8.0.0 (API 26) (emulator)
```

**Comment by Skylled** [comment]
I'm running into this one too.
Any suggestions for a workaround? Or a way to only set a TextField's initial text, instead of always asking the TextEditingController?

**Comment by cornelius-tan** [comment]
@Skylled Try initializing the controller with an initial text.
```dart
final TextEditingController _controller = new TextEditingController(text: 'INITIAL_TEXT_HERE');
```
source: https://docs.flutter.io/flutter/widgets/TextEditingController/TextEditingController.html

**Comment by Skylled** [comment]
@cornelius-tan That's exactly what I'm doing, but for whatever reason, in specific instances, it's resetting to the left when typing. I've found ways to work around the scenario instead of the problem.

**Comment by zoechi** [comment]
I think this is related. Not sure if it is exactly the same

With redux I have this code

```dart
final TextEditingController _textController = new TextEditingController();
...
_controller.messageComposeController.onTextChange
    .listen((text) => setState(() => _textController.text = text));
```
and
```dart
 new TextField(
     controller: _textController,
     onChanged: _controller.messageComposeController.textChanged,
```
but this way the cursor is always reset to position 0 after each keypress

**Comment by DrivEddy** [comment]
Same for me. With TextEditingController init text the curser is always at position 0

**Comment by etanxing** [comment]
Any workaround, guys? I’m facing the same issue.

**Comment by zoechi** [comment]
My workaround is to get selection and then set it again
```dart
              var cursorPos = _textController.selection;

              _textController.text = text ?? '';

              if (cursorPos.start > _textController.text.length) {
                cursorPos = new TextSelection.fromPosition(
                    new TextPosition(offset: _textController.text.length));
              }
              _textController.selection = cursorPos;
```

**Comment by sriharsha-dev** [comment]
Lets say you've the earlier value in **_username**  variable.
Then add the following as controller to the TextField

```
new TextField(
                        decoration: new InputDecoration(labelText: "Username"),
                        controller: new TextEditingController.fromValue(new TextEditingValue(text: _username,selection: new TextSelection.collapsed(offset: _username.length-1))),
                        onChanged: (val) => _username = val,
                      )
```

**Comment by b354c502** [comment]
None of these solutions worked for me. On android 8 it looks as though the cursor has gone to the end of the text, but as soon as the user starts typing, the cursor jumps to the beginning of the text.

**Comment by benhurott** [comment]
+1

**Comment by b354c502** [comment]
I traced this problem to re-instantiating the TextEditingController every time the build method is called. This is remedied by storing the TextEditingController as a variable of the class which extends State, and instantiating it only once in the initState method.

**Comment by ykphuah** [comment]
I have a different type of issue on iOS. I noticed that when I set the textController.text to something, when the textifeld is having focus, the cursor looked like it goes to the front of the text, but if I starts typing again, the "real" cursor is actually at the end. The new keys that I have typed get appended to the back and the cursor goes to the end.

So the behavior is as expected, just that the display of the cursor felt strange.

**Comment by XtremeTheory** [comment]
@sriharsha-dev WORKED FOR ME!  Thank you sir!

**Comment by marcglasberg** [comment]
Workaround:

    class TextEditingControllerWorkaroud extends TextEditingController {
      TextEditingControllerWorkaroud({String text}) : super(text: text);
    
      void setTextAndPosition(String newText, {int caretPosition}) {
        int offset = caretPosition != null ? caretPosition : newText.length;
        value = value.copyWith(
            text: newText,
            selection: TextSelection.collapsed(offset: offset),
            composing: TextRange.empty);
      }
    
      set text(String newText) {
        setTextAndPosition(newText);
      }
    }

Then:

    TextEditingController _controller = new TextEditingControllerWorkaroud();
    var tf = new TextField(controller: _controller);

You can use the setter, or you can also use the `setTextAndPosition` method.

**Comment by AlexandreRoba** [comment]
The work arround works  for me when typing letters at the end... But if the user move the cursor in  the textfield and add a character in the middle, the character is appended at the end.
This Scenario is used a lot when entering addresses. you start typing, you find the street on google places, you select and then you want to edit it with the proper house number.

**Comment by aalshaiba** [comment]
> This worked for me !
> Lets say you've the earlier value in **_username** variable.
> Then add the following as controller to the TextField
> 
> ```
> new TextField(
>                         decoration: new InputDecoration(labelText: "Username"),
>                         controller: new TextEditingController.fromValue(new TextEditingValue(text: _username,selection: new TextSelection.collapsed(offset: _username.length-1))),
>                         onChanged: (val) => _username = val,
>                       )
> ```

It worked with me, but I used _username.length without (-1)

**Comment by marcglasberg** [comment]
@AlexandreRoba If you mean my workaround, I have removed the setter. Please use method setTextAndPosition as needed.

**Comment by Firman95** [comment]
> @AlexandreRoba If you mean my workaround, I have removed the setter (see above). Please use method setTextAndPosition as needed.

How do you get correct offset if the user adds character in the middle of the string?

**Comment by akeating** [comment]
My use-case was to validate the TextField during input to modify the disabled status of a `Save` button and I've observed many of the usability concerns outlined in this issue.  

@ROTGP's solution appears to overcome all of my issues with the `TextField`. To reiterate; define your `TextEditingController` once in your widget. If you need to set an initial value, use the `initState` block.

Try something like this (modify to suit your needs):
```
class MyModal extends StatefulWidget {
  final initialTextFieldValue = 'fo';
  MyModal();
  State createState() => new MyModalState();
}

class MyModalState extends State<MyModal> {
  TextEditingController _textEditingController;
  bool _isValid;

  @override
  void initState() {
    _textEditingController = TextEditingController(text: widget.initialTextFieldValue);
    _isValid = _validate();
  }

  bool _validate() {
    return _textEditingController.text.length > 3;
  }

  Widget _build() {
    return TextField(
      controller: _textEditingController,
      textInputAction: TextInputAction.done,
      onChanged: (val) {
        setState(() {
          _isValid = _validate();
        });
      },
      keyboardType: TextInputType.multiline,
      maxLines: 2,
    )
  }

  @override
  Widget build(BuildContext context) {
    _onCancel() {
      Navigator.of(context).pop();
    }
    _onSave() {
      // maybe do another thing here
      Navigator.of(context).pop();
    }
    return Scaffold(
      appBar: AppBar(
        leading: FlatButton(
          onPressed: _onCancel,
          child: Icon(Icons.close),
          textColor: Colors.white,
        ),
        title: Text('My title'),
        centerTitle: true,
        actions: <Widget>[
          FlatButton(
            onPressed: _isValid ? _onSave : null,
            child: Text('Save', style: TextStyle(fontSize: 18.0),),
            textColor: Colors.white,
          ),
        ],
      ),
      body: _build()
    );
  }
}
```

**Comment by justinmc** [comment]
Is anyone still seeing this problem?  Running the original example given by @cornelius-tan works fine for me (iOS simulator).  There were a few recent cursor position PRs that improved things.

**Comment by krszwsk** [comment]
I encountered this issue when binding a `CupertinoTextField` to redux store value:

```dart
// some.state.dart
class SomeState {
  final String ownerField;

  SomeState({this.ownerField});

  SomeState copyWith({String ownerField}) =>
      SomeState(
          ownerField: ownerField ?? this.ownerField
      );

  factory SomeState.initial() => SomeState(ownerField: '');
}
```

```dart
// search_form.widget.dart

// ...

StoreConnector<AppState, HomeViewModel>(
  onDidChange: (store) {
    _ownerFieldController.text = store.ownerField;
  },
  converter: (store) => HomeViewModel.fromStore(store),
  builder: (_, viewModel) {
    this.viewModel = viewModel;
    return render(viewModel);
  },
);

// ...

CupertinoTextField(
  onSubmitted: _focusRepoField,
  textInputAction: TextInputAction.next,
  onChanged: viewModel.setOwnerField,
  controller: _ownerFieldController,
),
// ...
```

For now I fixed it by updating the field text only if there was an external value change:

```dart
// StoreConnector argument
      onDidChange: (store) {
        if (_ownerFieldController.text != store.ownerField)
          _ownerFieldController.text = store.ownerField;
      },
```

I feel like I may be doing something incorrectly though. If so, please point that out. Thanks 🙂.

**Comment by shabab477** [comment]
I am using a hacky solution for setting the cursor to the end if it is below a certain length. 

```
void _textChangeListener() {
    if (_textEditingController.text.length <= 3) {
      _textEditingController.text = "+880";
      _textEditingController.selection = TextSelection.collapsed(offset: "+880".length)..end;

      return;
    }
}
```

**Comment by jaumard** [comment]
Shouldn't be closed, it's still a bug imho `TextEditingController(text: 'test')` should put the cursor at the end of the field, not start

**Comment by jaumard** [comment]
I'm using stable 1.0.0 on my side (on iOS simulator) and have the problem describe. Maybe on another channel it works better :)

**Comment by zoechi** [comment]
The comment is from 

> justinmc commented 23 days ago

If you run `flutter doctor -v` you'll see that your Flutter version is 9 weeks old. So that doesn't count. `dev` and `master` already contain the fixes.

**Comment by hjJunior** [comment]
I have changed branch master and I keeps not working, maybe I missing something, could u help please?

```
[✓] Flutter (Channel master, v1.2.1-pre.61, on Mac OS X 10.13.6 17G65, locale pt-BR)
    • Flutter version 1.2.1-pre.61 at /Users/helio.s.junior/development/flutter
    • Framework revision 4e235c0cd8 (41 minutes ago), 2019-02-01 18:01:43 +0100
    • Engine revision 15f2b92cce
    • Dart version 2.1.1 (build 2.1.1-dev.3.2 71bee8f05e)

[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /Users/helio.s.junior/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1136-b06)
    • All Android licenses accepted.

[✓] iOS toolchain - develop for iOS devices (Xcode 10.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 10.1, Build version 10B61
    • ios-deploy 1.9.4
    • CocoaPods version 1.5.3

[✓] Android Studio (version 3.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 31.3.1
    • Dart plugin version 181.5656
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1136-b06)

[✓] Connected device (1 available)
    • iPhone XR • 727F93C0-A7A8-4139-8C43-F4846DC3D7BC • ios • iOS 12.1 (simulator)
```

**Comment by justinmc** [comment]
@hjJunior I just tried again with master and the example given at the top, and it is working for me.  What exactly are you running that doesn't work?

**Comment by temirfe** [comment]
so what is the final solution?

**Comment by hjJunior** [comment]
Just move to another channel `dev` or `master`, both already contain the fixes

**Comment by opctim** [comment]
> This worked for me !
> Lets say you've the earlier value in **_username** variable.
> Then add the following as controller to the TextField
> 
> ```
> new TextField(
>                         decoration: new InputDecoration(labelText: "Username"),
>                         controller: new TextEditingController.fromValue(new TextEditingValue(text: _username,selection: new TextSelection.collapsed(offset: _username.length-1))),
>                         onChanged: (val) => _username = val,
>                       )
> ```

Nice! Works!

**Comment by temirfe** [comment]
@sriharsha-dev suggested works but when you place cursor somewhere middle and start typing the cursor goes to the end.

**Comment by shoxter** [comment]
I'm still experiencing this on 1.2.0. None of the work arounds work for me either.

**Comment by zoechi** [comment]
@shoxter I don't think this is planned to change. You can set the cursor position yourself like shown above.

**Comment by andyduke** [comment]
Now it still does not work correctly. I can not put the cursor to the very end, it moves to the beginning.

**Comment by shoxter** [comment]
None of the implementations above work for me, unfortunately.

**Comment by zoechi** [comment]
I'd suggest to ask on StackOverflow

**Comment by intranetsfr** [comment]
`
  TextEditingController dobController = new TextEditingController(text: "");


  @override
  void initState() {
    super.initState();
    dobController.addListener(_setDobValue);
  }
  void _setDobValue() {
    print(dobController.text.length);
    var cursorPos = dobController.selection;
    if (dobController.text.length == 2) {
      dobController.text = dobController.text + "/";
    }
    if (dobController.text.length == 4) {
      dobController.text = dobController.text + "/";
    }
    cursorPos = new TextSelection(baseOffset: dobController.text.length, extentOffset: dobController.text.length);
    dobController.selection = cursorPos;
  }
`
That's ok for me

**Comment by vware** [comment]
Same issue here, tried to get around it with a FocusNode and issuing a onChanged callback only when focus was lost - however, TextField doesn't lose focus when one taps a switch or assorted other UI elements. Also, the keyboard doesn't go away when clicking a Switch... why not? :(
I don't see any changes in behavior on 1.2.0 - so what is the best solution for this? Just not call setState and/orkeep the TextEditingController instances instead of rebuilding them on each key press in the TextField?

**Comment by h-soran** [comment]
Fix for the CupertinoTextField: this moves the cursor to the end.
...
controller: _controller,
onChanged: (value) {
                _controller.text = value;
                _controller.selection = TextSelection.collapsed(offset: value.length);
            }
...

**Comment by atereshkov** [comment]
Lol still issue within 2 years :D

**Comment by thosakwe** [comment]
Simple change to @marcglasberg 's answer that makes everything "just work" for me:

```dart
class SelectingTextEditingController extends TextEditingController {
  SelectingTextEditingController({String text}) : super(text: text) {
    if (text != null) setTextAndPosition(text);
  }

  void setTextAndPosition(String newText, {int caretPosition}) {
    int offset = caretPosition != null ? caretPosition : newText.length;
    value = value.copyWith(
        text: newText,
        selection: TextSelection.collapsed(offset: offset),
        composing: TextRange.empty);
  }
}
```

**Comment by h-soran** [comment]
No need a workaround, it seems.

The problem occurs when we use a TextController and set TextField's value in onChanged function; because TextController does the setting automatically.  If we reserve the onChanged function to do other things than manipulating the field's value it works as expected for both CupertineTextField and TextField.

CupertinoTextField(
        keyboardType: TextInputType.emailAddress,
        controller: _emailController,
        onChanged: (value) {
          print('value: $value');
          print('emailController: ${_emailController.text}');
        });

**Comment by thosakwe** [comment]
@h-soran: My above workaround is for cases where you are creating a new TextController in each build, i.e. in a StatelessWidget. AFAIK there is no way to pass text to a TextField without a TextController.

Sure, we should technically be using StatefulWidgets, but still.

**Comment by mijogu** [comment]
I was dealing with this cursor issue as well with both TextField and TextFormField inputs. Oddly enough, the cursor worked as expected for my password fields with `obscureText: true;`

Just updated to Flutter v1.3.10 and the issue is fixed!

**Comment by gagaJay** [comment]
> I have a different type of issue on iOS. I noticed that when I set the textController.text to something, when the textifeld is having focus, the cursor looked like it goes to the front of the text, but if I starts typing again, the "real" cursor is actually at the end. The new keys that I have typed get appended to the back and the cursor goes to the end.
> 
> So the behavior is as expected, just that the display of the cursor felt strange.

We’re you able to fix the issue? @ykphuah

**Comment by h-soran** [comment]
> > I have a different type of issue on iOS. I noticed that when I set the textController.text to something, when the textifeld is having focus, the cursor looked like it goes to the front of the text, but if I starts typing again, the "real" cursor is actually at the end. The new keys that I have typed get appended to the back and the cursor goes to the end.
> > So the behavior is as expected, just that the display of the cursor felt strange.
> 
> We’re you able to fix the issue? @ykphuah

See my answer above; it solves the cursor problem! 
To initialize a textfield use the constructor in the init function. e.g. Textcontroller _emailController = TextController('name@domain.com');
Avoid doing this in the onChange function: _emailController.text = value
The cursor issue occurs if you do that.

**Comment by vijaykanta** [comment]
I got it around @zoechi answer to set selection to the length of the text, however my only problem is the name of the title itself and the examples given. It's not the cursor, it's the 'CARET' that blinks on screen.

**Comment by annezhou920** [comment]
piggying back on what @h-soran is saying...the cursor issue is because you are directly assigning the text value in the `onChanged` property

i followed this guide from their docs to properly handle changes to a text field and was able to avoid the cursor issue: https://flutter.dev/docs/cookbook/forms/text-field-changes

**Comment by sgon00** [comment]
Hi, by resetting `_controller.selection`, it works. But there is a small visual problem, you can see that visually the cursor first jumps to the beginning of the text and then jump to the end. Do anyone know how to avoid this unexpected cursor jump? Should I open a new bug for this since this issue is closed? Thanks a lot.

**Comment by justinmc** [comment]
@sgon00 Yes, that sounds like a separate issue would be helpful if you can provide some code to reproduce it.  Be sure to reference this issue so that it shows up here.

**Comment by sgon00** [comment]
@justinmc thank you very much for the quick reply. I actually found two new problems based on this issue. I filed another bug: https://github.com/flutter/flutter/issues/32209.

**Comment by Firman95** [comment]
Form me, it works like this, the cursor is right where it's supposed to be.

  ```
final TextEditingController _controller = TextEditingController();

  Widget _textField(BuildContext context, String stringData) {
    _controller.value = TextEditingValue(
        text: stringData, selection: _textInputController.selection);

    return TextField(
      controller: _textInputController,
      onChanged: _updateStringData,
      maxLines: 4,
    );
  }
```

**Comment by sgon00** [comment]
I think there are some people who commented on this issue do not know what this issue really is. This issue is about how to change TextField's text programmatically by using TextEditController. And the reason why the cursor moves to the beginning is because the cursor position is always set to -1 when the value is modified programmatically by TextEditingController. I have filed a new issue #32209 for this.

**Comment by tewshi** [comment]
If you're replacing the text with a new value (suggestion)
```
_inputCtrl.value = TextEditingValue(
  text: suggestion,
  selection: TextSelection.collapsed(offset: suggestion.length),
);
```

**Comment by VickySalunkhe** [comment]
I am facing almost same issue.
I am fetching data from server
then setting the text via a function using TextEditingController.

whenever i try enter a space the cursor doesn't move and as soon as i enter a character other than space cursor suddenly move to the end of that character which seems to be odd behavior from user point of view.

**Comment by justinmc** [comment]
@VickySalunkhe You're looking for this issue, which is still waiting for a fix: https://github.com/flutter/flutter/issues/30975

**Comment by bawelter** [comment]
I found that the issue is actually the TextEditingController is being initialized / changed because the widget tree is being rebuilt.  Move the initialization of the controller outside of the functions building the textfield widget and it should resolve the issue.

**Comment by jjay225** [comment]
As what @bawelter and others said: In my `Widget build(BuildContext context) { }` 
code section I was setting the text property of my TextEditingController for this edit screen of mine.  The TextEditingControllers were being instantiated in the start of the state class: 

`class TestDetailState extends State<TestDetail>
  {
  TextEditingController itemNameController = TextEditingController();
 }
`
As soon as I moved the setting of the text property of the TextEditingController out of the build method and put it say below the instantiate line the cursor correctly goes to the end of the line of text in the text box.

**Comment by lifenautjoe** [comment]
Is this fixed on master or dev? Still broken in `stable`

**Comment by justinmc** [comment]
@lifenautjoe I tried the reproduction steps given by the original description above and I don't see the problem on master.

**Comment by lifenautjoe** [comment]
Alright, in stable I had to do 
```dart
_controller.text = newValue;
_controller.selection = TextSelection.collapsed(offset: _textController.text.length)
```

but I still see the cursor going to beginning and then end in a blink.

Supposedly `_controller.value(newValue)` does this behind the scenes but doesn't work here.

**Comment by bawelter** [comment]
There is no bug just error in our coding.  Refer to my last comment above as the likely issue.

**Comment by hernandoh** [comment]
@lifenautjoe this is the great solution, including it in onchange of textfield

```
onChanged: (password) {
  passwordControl.selection = new TextSelection(baseOffset: password.length, extentOffset: password.length);
}
```

**Comment by sandeepgurram** [comment]
> @Skylled Try initializing the controller with an initial text.
> 
> ```dart
> final TextEditingController _controller = new TextEditingController(text: 'INITIAL_TEXT_HERE');
> ```
> 
> source: https://docs.flutter.io/flutter/widgets/TextEditingController/TextEditingController.html

this solved my issue. In my case am using onChange callback and was also notifieres  to update the textfields. Setting default value every time in build() method solved my issue

**Comment by karnadii** [comment]
i have a virtual keyboard, i want to add a character at the start, the middle and the end of a line, my solution is below
```dart
  onVirtualKeyboardPressed(VirtualKeyboardKey key) {
    String text = textCtrl.text;
    var cursorPos = textCtrl.selection;
    if (cursorPos.end == 0 || cursorPos.end == -1) {
      text = key.text + text;
    } else if (cursorPos.end == text.length) {
      text = text + key.text;
    } else {
      text = text.substring(0, textCtrl.selection.end) +
          key.text +
          text.substring(textCtrl.selection.end, textCtrl.text.length);
    }
    textCtrl = TextEditingController(text: text)
      ..selection = TextSelection.collapsed(
          offset: cursorPos.end == -1 ? cursorPos.end + 2 : cursorPos.end + 1);

    onTextFieldValueChanged(text);
  }
```
with my code above, the cursor won't move to the start or moving to the end of text after the cursor moving to the start.

**Comment by ccadieux** [comment]
`controller: new TextEditingController.fromValue(new TextEditingValue(text: _username,selection: new TextSelection.collapsed(offset: _username.length))),`

I tried this method and it works but has a minor defect. On iOS you can't trigger the text selection toolbar in an empty textfield with this method. Setting up the controller in initState doesn't have this issue.

**Comment by VickySalunkhe** [comment]
I solved my issue by using this

this will set your cursor at the end.

    final val = TextSelection.collapsed(offset: _textTEC.text.length); 
     _textTEC.selection = val;

**Comment by burkard** [comment]
> 
> 
> I solved my issue by using this
> 
> this will set your cursor at the end.
> 
> ```
> final val = TextSelection.collapsed(offset: _textTEC.text.length); 
>  _textTEC.selection = val;
> ```

This solved the problem. And only 2 hours after I found the problem! 🙉

**Comment by jifalops** [comment]
I think a TextInputFormatter should be used instead of setting controller.text when the text changes. See https://stackoverflow.com/a/49239762/884522.

**Comment by bernaferrari** [comment]
Why is this issue still closed? If it still present (and it is) it should be open.

**Comment by justinmc** [comment]
@bernaferrari What's the exact problem you're seeing?  If you run the code in the original description do you see the same problem?

**Comment by bernaferrari** [comment]
Yes, I needed to add the code from https://github.com/flutter/flutter/issues/11416#issuecomment-405792611 to solve my issue.

**Comment by cgs** [comment]
Another workaround is to save the cursor position as an attribute then set the selection property on the controller. In my case I need to do it in didChangeDependencies but ymmv. If you edit somewhere in the middle of the text, it will retain that position.
```
class _MyWidgetWithTextFieldState extends State<MyWidgetWithTextField> {
  final TextEditingController _controller = TextEditingController();
  int _cursorOffset;

  @override
  void didChangeDependencies() {
    _controller.selection = TextSelection.collapsed(
        offset: _cursorOffset ?? _controller.text.length);
  }

  // this is called in the TextField's onChange:
  void _saveTextField(BuildContext context, String n) {
    _cursorOffset = _controller.selection.baseOffset;
    // save to DB here...
  }

// ...
}
```

**Comment by iKK001** [comment]
This should be the preferred answer ! Thank you very much, @annezhou920 , for your answer !!
The textController.listener was doing the wonder work !!

> piggying back on what @h-soran is saying...the cursor issue is because you are directly assigning the text value in the `onChanged` property
> 
> i followed this guide from their docs to properly handle changes to a text field and was able to avoid the cursor issue: https://flutter.dev/docs/cookbook/forms/text-field-changes

**Comment by devibrahimkarahan** [comment]
Solution with extension
```
extension TextControllerSetTextAndSelection on TextEditingController {
  void setTextAndSelection(String text, [int selectionOffset]) {
    selectionOffset ??= text.length;
    this.text = text;
    this.selection = TextSelection.collapsed(offset: selectionOffset);
  }
}
```
Usage
```
controller.setTextAndSelection("new text with extension method");
```

**Comment by demsey2** [comment]
I got the following error with that extension

```
A member named 'setTextAndSelection' is defined in extensions 'TextControllerSetTextAndSelection' and 'TextControllerSetTextAndSelection' and neither is more specific.
Try using an extension override to specify the extension you want to to be chosen.dart(ambiguous_extension_member_access)
```

do you know how to fix that?

**Comment by devibrahimkarahan** [comment]
I dont know, but this issue is about your error. [#39262](https://github.com/dart-lang/sdk/issues/39262)

**Comment by itsgauravjain22** [comment]
Yep, its a problem with flutter. Coming from react-native environment, the text controller thing is a messed up concept. For each text input having initial value, you need to have a text controller for each input field on the screen. What's more to the pain is the cursor location doesn't maintain its position. 

I think the flutter team should consider adding `value` argument in the text field widget. That would ease the pain and makes the text editing simpler and unified. 

If that's not possible, kindly change the default behavior to have the same cursor position where it was, when a user is editing the value instead of moving the cursor at the beginning. It's pretty much required for most of the input scenarios and expected from the flutter team to be resolved.

**Comment by hamza39460** [comment]
> Alright, in stable I had to do
> 
> ```dart
> _controller.text = newValue;
> _controller.selection = TextSelection.collapsed(offset: _controller.text.length)
> ```
> 
> but I still see the cursor going to beginning and then end in a blink.
> 
> Supposedly `_controller.value(newValue)` does this behind the scenes but doesn't work here.

This worked for me ♥

**Comment by QuentinSc** [comment]
> Alright, in stable I had to do
> 
> ```dart
> _controller.text = newValue;
> _controller.selection = TextSelection.collapsed(offset: _controller.text.length)
> ```
> 
> but I still see the cursor going to beginning and then end in a blink.
> 
> Supposedly `_controller.value(newValue)` does this behind the scenes but doesn't work here.

Same
Flutter V1.17. Become crazy with this 0_o

**Comment by rodion-m** [comment]
```
textController.selection = TextSelection.fromPosition(TextPosition(offset: textController.text.length))
```
Still working on version 1.19.0-4.0.pre.

**Comment by PhilTran1994** [comment]
> > This worked for me !
> > Lets say you've the earlier value in **_username** variable.
> > Then add the following as controller to the TextField
> > ```
> > new TextField(
> >                         decoration: new InputDecoration(labelText: "Username"),
> >                         controller: new TextEditingController.fromValue(new TextEditingValue(text: _username,selection: new TextSelection.collapsed(offset: _username.length-1))),
> >                         onChanged: (val) => _username = val,
> >                       )
> > ```
> 
> It worked with me, but I used _username.length without (-1)

This is worked for me also, some how new update recently textcontroller not doesn't work really well

**Comment by omidraha** [comment]
Related [SO](https://stackoverflow.com/questions/56851701/how-to-set-cursor-position-at-the-end-of-the-value-in-flutter-in-textfield).

**Comment by denzilferreira** [comment]
> ```
> class TextEditingControllerWithCursorPosition extends TextEditingController {
>   TextEditingControllerWithCursorPosition({String text}): super(text: text);
> 
>   @override
>   set text(String newText) {
>     value = value.copyWith(
>       text: newText,
>       selection: value.selection,
>       composing: TextRange.empty,
>     );
>   }
> }
> ```

This is the right answer. It acts like the TextEditingController!

**Comment by campovski** [comment]
@sriharsha-dev works, however keep just `_username.length` instead of the option with `-1`!

**Comment by cdmunoz** [comment]
Finally I could make something that works in my scenario using software keyboard and hardware keyboard for emulators:

First within `initState`:

```dart
_myController.value = _myController.value.copyWith(
            text: myText,
            selection: TextSelection(
                baseOffset: myText.length, extentOffset: myText.length));
```

Then set the controller to the text field

**Comment by chinitadelrey** [comment]
> > ```
> > class TextEditingControllerWithCursorPosition extends TextEditingController {
> >   TextEditingControllerWithCursorPosition({String text}): super(text: text);
> > 
> >   @override
> >   set text(String newText) {
> >     value = value.copyWith(
> >       text: newText,
> >       selection: value.selection,
> >       composing: TextRange.empty,
> >     );
> >   }
> > }
> > ```
> 
> 
> This is the right answer. It acts like the TextEditingController!

Correct me if I am wrong, but this solution would not work if the newText is not the same length as the value. Replacing 

> >       selection: value.selection,

with 
>>      selection: TextSelection.collapsed(offset: newText.length),

does the trick. When you initialize your custom text controller, you can do it like this: 

>  TextEditingController searchController = new TextEditingControllerWithCursorPosition();
